test/Microsoft.AspNetCore.Mvc.Core.Test/ModelBinding/Internal/ValidationStackTest.cs

```diff
@@ -0,0 +1,127 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using Microsoft.AspNetCore.Mvc.ModelBinding.Internal;
+using Xunit;
+
+namespace Microsoft.AspNetCore.Mvc.Core.Test.ModelBinding.Internal
+{
+    public class ValidationStackTest
+    {
+        private class TestValidationStack : ValidationStack
+        {
+            public bool Contains(object model)
+            {
+                if (_hashSet != null)
+                {
+                    return _hashSet.Contains(model);
+                }
+                else
+                {
+                    return _list.Contains(model);
+                }
+            }
+
+            public bool UsingHashSet()
+            {
+                return _hashSet != null;
+            }
+        }
+
+        [Theory]
+        [InlineData(0)]
+        [InlineData(23)]
```

Comment:
The `[Theory]`s test the initial `Push()` and `Push()` after the transition to a `HashSet()`. Suggest also testing something in the middle for each e.g. `[InlineData(5)]`.